crates/bin/src/initialise.rs

```diff
@@ -0,0 +1,129 @@
+use std::{env::var_os, fs, path::PathBuf};
+
+use binstalk::errors::BinstallError;
+use binstalk_manifests::{cargo_config::Config as CargoConfig, crates_manifests::Manifests};
+use home::cargo_home;
+use miette::{bail, Result, WrapErr};
+use tempfile::TempDir;
+use tracing::{debug, info};
+
+use crate::{args::Args, ui::confirm_sync};
+
+pub(crate) struct Init {
+    pub(crate) cargo_config: CargoConfig,
+    pub(crate) settings: crate::settings::Settings,
+    pub(crate) cargo_root: PathBuf,
+    pub(crate) install_path: PathBuf,
+    pub(crate) manifests: Option<Manifests>,
+    pub(crate) temp_dir: TempDir,
+}
+
+pub(crate) fn initialise(args: &Args) -> Result<Init> {
+    let (cargo_config, cargo_home) = if let Ok(home) = cargo_home() {
+        (
+            Some(CargoConfig::load_from_path(home.join("config.toml"))?),
+            Some(home),
+        )
+    } else {
+        (None, None)
+    };
+
+    let cargo_root = if let Some(p) = &args.root {
+        debug!(path=?p, "install root from --root");
+        p.into()
+    } else if let Some(p) = var_os("CARGO_INSTALL_ROOT").map(PathBuf::from) {
+        debug!(path=?p, "install root from CARGO_INSTALL_ROOT");
+        p
+    } else if let Some(p) = cargo_config
+        .as_ref()
+        .and_then(|config| config.install.clone().and_then(|install| install.root))
+    {
+        debug!(path=?p, "install root from cargo config");
+        p
+    } else if let Some(p) = &cargo_home {
+        debug!(path=?p, "install root from cargo home");
+        p.into()
+    } else if let Some(p) = dirs::executable_dir() {
+        debug!(path=?p, "install root from executable dir");
+        p
+    } else {
+        bail!("No install root found, provide one with --root or CARGO_INSTALL_ROOT");
+    };
+
+    let cargo_config = if let Some(config) = cargo_config {
+        config
+    } else {
+        CargoConfig::load_from_path(cargo_root.join("config.toml"))?
+    };
+
+    let settings_path = args.settings.clone().unwrap_or(
+        cargo_home
+            .as_ref()
+            .unwrap_or(&cargo_root)
+            .join("binstall.toml"),
+    );
+    let mut settings = crate::settings::load(args.settings.is_some(), &settings_path)?;
```

Comment:
If we ignore the error when args not set, then we might still fail at installation, e.g. the settings is occupied by a directory/special file, or no permissions

Reply:
Ah ok I've got it, having a default if args.settings already exists so that we can merge it and continue

Reply:
And the default value on error won't trigger any consent or saving, so it'd be ok